tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -197,65 +166,44 @@ build_compressed_scan_pathkeys(SortInfo *sort_info, PlannerInfo *root, List *chu
 	 */
 	if (info->num_segmentby_columns > 0)
 	{
-		Bitmapset *segmentby_columns = bms_copy(info->chunk_segmentby_ri);
+		TimescaleDBPrivate *compressed_fdw_private =
+			(TimescaleDBPrivate *) info->compressed_rel->fdw_private;
+		/*
+		 * We don't need any sorting for the segmentby columns that are equated
+		 * to a constant. The respecitve constant ECs are excluded from
+		 * canonical pathkeys, so we won't see these columns here. Count them as
+		 * seen from the start, so that we arrive at the proper counts of seen
+		 * segmentby columns in the end.
+		 */
+		Bitmapset *segmentby_columns = bms_copy(info->chunk_const_segmentby);
 		ListCell *lc;
-		char *column_name;
-		Oid sortop;
-		Oid opfamily, opcintype;
-		int16 strategy;
-
 		for (lc = list_head(chunk_pathkeys);
 			 lc != NULL && bms_num_members(segmentby_columns) < info->num_segmentby_columns;
 			 lc = lnext(chunk_pathkeys, lc))
 		{
 			PathKey *pk = lfirst(lc);
-			var = (Var *) find_em_expr_for_rel(pk->pk_eclass, info->chunk_rel);
-
-			if (var == NULL || !IsA(var, Var))
-				/* this should not happen because we validated the pathkeys when creating the path
-				 */
-				elog(ERROR, "Invalid pathkey for compressed scan");
-
-			/* not a segmentby column, rest of pathkeys should be handled by compress_orderby */
-			if (!bms_is_member(var->varattno, info->chunk_segmentby_attnos))
-				break;
-
-			/* skip duplicate references */
-			if (bms_is_member(var->varattno, segmentby_columns))
-				continue;
-
-			column_name = get_attname(info->chunk_rte->relid, var->varattno, false);
-			segmentby_columns = bms_add_member(segmentby_columns, var->varattno);
-			varattno = get_attnum(info->compressed_rte->relid, column_name);
-			var = makeVar(info->compressed_rel->relid,
-						  varattno,
-						  var->vartype,
-						  var->vartypmod,
-						  var->varcollid,
-						  0);
-
-			sortop =
-				get_opfamily_member(pk->pk_opfamily, var->vartype, var->vartype, pk->pk_strategy);
-			if (!get_ordering_op_properties(sortop, &opfamily, &opcintype, &strategy))
+			EquivalenceMember *compressed_em = NULL;
+			ListCell *ec_em_pair_cell;
+			foreach (ec_em_pair_cell, compressed_fdw_private->compressed_ec_em_pairs)
 			{
-				if (type_is_enum(var->vartype))
-				{
-					sortop = get_opfamily_member(pk->pk_opfamily,
-												 ANYENUMOID,
-												 ANYENUMOID,
-												 pk->pk_strategy);
-				}
-				else
+				List *pair = lfirst(ec_em_pair_cell);
+				if (linitial(pair) == pk->pk_eclass)
 				{
-					elog(ERROR, "sort operator lookup failed for column \"%s\"", column_name);
+					compressed_em = lsecond(pair);
+					break;
 				}
 			}
-			pk = make_pathkey_from_compressed(root,
-											  info->compressed_rel->relid,
-											  (Expr *) var,
-											  sortop,
-											  pk->pk_nulls_first);
+
+			if (compressed_em == NULL)
+			{
+				/* Not a segmentby column, rest of pathkeys should be handled by compress_orderby */
```

Comment:
Is it possible to test this condition?

Reply:
I double-checked and seems like it is redundant with the loop condition, so I changed it to an assertion.